colbert/indexing/codecs/residual_embeddings.py

```diff
@@ -29,28 +28,50 @@ def load_chunks(cls, index_path, chunk_idxs, num_embeddings):
 
         dim, nbits = get_dim_and_nbits(index_path)
 
-        codes = torch.empty(num_embeddings, dtype=torch.int32)
-        residuals = torch.empty(num_embeddings, dim // 8 * nbits, dtype=torch.uint8)
+        print_message("#> Loading codes and residuals...")
 
-        codes_offset = 0
+        if len(chunk_idxs) == 1: # Memory mapping is enabled for single-file indexes
+            residuals_path = os.path.join(index_path, f'0.residuals.pt')
+            codes_path = os.path.join(index_path, f'0.codes.pt')
 
-        print_message("#> Loading codes and residuals...")
+            codes_size = get_codes_size(index_path, 0)
+            storage = torch.IntStorage.from_file(filename=codes_path, shared=True, size=codes_size + 80)
+            codes = torch.IntTensor(storage)[80:]
 
-        for chunk_idx in tqdm.tqdm(chunk_idxs):
-            chunk = cls.load(index_path, chunk_idx)
+            residuals_size, codes_size, packed_dim = get_residuals_size(index_path, 0)
+            storage = torch.ByteStorage.from_file(filename=residuals_path, shared=True, size=residuals_size + 320)
+            ret = torch.ByteTensor(storage)
+            ret = ret[320:]
+            unflatten = torch.nn.Unflatten(0, (codes_size, packed_dim))
+            residuals = unflatten(ret)
 
-            codes_endpos = codes_offset + chunk.codes.size(0)
+            print("Loaded a single memory-mapped index.")
+        else:
+            codes = torch.empty(num_embeddings, dtype=torch.int32)
+            residuals = torch.empty(num_embeddings, dim // 8 * nbits, dtype=torch.uint8)
 
-            # Copy the values over to the allocated space
-            codes[codes_offset:codes_endpos] = chunk.codes
-            residuals[codes_offset:codes_endpos] = chunk.residuals
+            codes_offset = 0
 
-            codes_offset = codes_endpos
+            for chunk_idx in tqdm.tqdm(chunk_idxs):
+                chunk = cls.load(index_path, chunk_idx)
 
-        # codes, residuals = codes.cuda(), residuals.cuda()  # FIXME: REMOVE THIS LINE!
+                codes_endpos = codes_offset + chunk.codes.size(0)
+
+                # Copy the values over to the allocated space
+                codes[codes_offset:codes_endpos] = chunk.codes
+                residuals[codes_offset:codes_endpos] = chunk.residuals
+
+                codes_offset = codes_endpos
 
         return cls(codes, residuals)
 
+    @classmethod
+    def load_from_single_index(self):
+        path = "/future/u/udingank/ColBERT/experiments/default/indexes/msmarco_single.nbits=2.latest/"
```

Comment:
Replace the hard-coded path

Reply:
load_from_single_index() is actually a stale function, going to get rid of it completely.